relay-server/src/services/project_cache.rs

```diff
@@ -898,19 +884,47 @@ impl ProjectCacheBroker {
 
     fn handle_flush_buckets(&mut self, message: FlushBuckets) {
         let metric_outcomes = self.metric_outcomes.clone();
+        let outcome_aggregator = self.services.outcome_aggregator.clone();
+        let aggregator = self.services.aggregator.clone();
+        let project_cache = self.services.project_cache.clone();
 
+        let mut no_project = 0;
         let mut scoped_buckets = BTreeMap::new();
         for (project_key, buckets) in message.buckets {
             let project = self.get_or_create_project(project_key);
-            if let Some((scoping, b)) = project.check_buckets(&metric_outcomes, buckets) {
-                scoped_buckets.insert(scoping, b);
+            match project.check_buckets(&metric_outcomes, &outcome_aggregator, buckets) {
+                CheckBuckets::NoProject(buckets) => {
+                    no_project += 1;
+                    // Schedule an update for the project just in case.
+                    project.prefetch(project_cache.clone(), false);
+                    project.merge_buckets(&aggregator, buckets, BucketSource::Internal);
```

Comment:
The only real overlap is the prefetch of the project, the merge buckets call is different and at best this could be a freestanding function due to borrowing rules passing in all arguments.